### PR TITLE
Static Typing / Refactoring

### LINKED ISSUE #5: Add static typing

### PR DESCRIPTION
**Pull request by marlonmueller**
- [ ] Closes #5 

Future PR
- [ ] Testing!
- [ ] ReadMe(?)

### CODE REVIEW

File: .github/workflows/continuous-integration.yml [hunk 1]
```diff
@@ -33,8 +33,8 @@ jobs:
       - name: Install missed dependencies
         run: poetry install
 
-      - name: Format with black
-        run: poetry run black . --check --diff
+      #- name: Format with black
+      #  run: poetry run black . --check --diff
```
Comment (dostuffthatmatters) [inline .github/workflows/continuous-integration.yml:37]: @MarlonMueller Why do you need to format it in the CI? The CI doesn't add new commits.

Comment (MarlonMueller) [reply .github/workflows/continuous-integration.yml:37]: It just enforces that the files are formatted in the first place.

Comment (dostuffthatmatters) [reply .github/workflows/continuous-integration.yml:37]: Aah! Thank you, that's a good idea 🙌

File: pyproject.toml [hunk 2]
```diff
@@ -1,26 +1,31 @@
...
-description = ""
+version = "0.5.0"
+description = "Download .map, .vmr and .mod files from ccycle.gps.caltech.edu"
 authors = [
-    "Florian Dietrich <flo.dietrich@tum.de>",
+    "Marlon Mueller <marlon.mueller@tum.de>",
+    "Moritz Makowski <moritz.makowski@tum.de>",
     "Xinxu Zhao <xinxu.zhao@tum.de>",
-    "Moritz Makowski <moritz.makowski@tum.de>"
+    "Florian Dietrich <flo.dietrich@tum.de>"
```
Comment (dostuffthatmatters) [inline pyproject.toml:9]: You can safely remove Florian and Xinxu in here.

File: pyproject.toml [hunk 3]
```diff
@@ -1,26 +1,31 @@
 [tool.poetry]
 name = "download-vertical-profiles"
-version = "0.4.0"
-description = ""
+version = "0.5.0"
+description = "Download .map, .vmr and .mod files from ccycle.gps.caltech.edu"
```
Comment (dostuffthatmatters) [inline pyproject.toml:4]: Could you rephrase this line as "Download vertical profiles (GGG2014 and GGG2020) from ccycle.gps.caltech.edu"?

File: src/custom_types/query.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    from_date: Date
+    to_date: Date
+    location: QueryLocation
+
+    def to_json(self) -> dict[str, Any]:
+        return {
+            "location": self.location.slug(),
+            "from_date": self.from_date,
+            "to_date": self.to_date,
+        }
```
Comment (dostuffthatmatters) [inline src/custom_types/query.py:47]: Generating the slug in here is very nice! But every pydantic model has the built-in function `.to_dict()`. Could you rename this to something more explicit?

File: src/custom_types/query.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    }.
+    """
+
+    lat: int
+    lon: int
+
+    class Config:
+        frozen = True
+
+    def slug(self, verbose: bool = False) -> str:
```
Comment (dostuffthatmatters) [inline src/custom_types/query.py:21]: Could you add an example slug as a comment (for verbose=true/false)?

File: src/procedures/location_data.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+
+def get_sensor_data(config: GitHubConfig) -> dict[SensorId, Sensor]:
+    """
+    Requests sensors.json from GitHub.
+    Returns a dictionary that maps a sensors_id to its sensor.
+    """
+    url = os.path.join(config.data_dir, "sensors.json")
+    response = utils.network.request_url(url, config.username, config.token)
+    return {sensor_id: Sensor(**data) for sensor_id, data in json.loads(response).items()}
```
Comment (dostuffthatmatters) [inline src/procedures/location_data.py:31]: Very elegant! 🙌

File: src/procedures/query.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+            (
+                "tu",
+                query.from_date,
+                to_date,
+                str(query.location.lat),
+                str(query.location.lon),
+                config.email,
+            )
+        ).encode("utf-8")
+    ) as file_:
```
Comment (dostuffthatmatters) [inline src/procedures/query.py:93]: Very good decision on having this as an in-memory file!

File: src/utils/network.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+def request_url(
+    url: str,
+    username: str,
+    password: str,
+    timeout: int = 10,
+) -> str:
+    """Sends a request and returns the content of the response, in unicode."""
+    response = requests.get(url, auth=(username, password), timeout=timeout)
+    response.raise_for_status()
```
Comment (dostuffthatmatters) [inline src/utils/network.py:12]: I didn't know this function yet. Good call!

File: tests/test_sensor_set.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        },
+        QueryLocation(lat=11, lon=11): {
+            "10000102": {"D"},
+            "10000103": {"D"},
+            "10000104": {"A", "D"},
+            "10000105": {"D"},
+        },
+    }
+
+    assert daily_sensor_sets == get_daily_sensor_sets(config, locations, sensors)
```
Comment (dostuffthatmatters) [inline tests/test_sensor_set.py:61]: Very clean!

**Review by dostuffthatmatters — APPROVED** [review]
Great improvements and nicely packaged. Well done!

In added a few very small change requests.

File: src/custom_types/query.py (hunk 4 above)
Comment (MarlonMueller) [reply src/custom_types/query.py:47]: Good call (could also be misleading as location is not `{lat:..., lon:...} `but, e.g., `48N012E`). If we leave it like this (only used for the reports) maybe something like `.json_slug()`?

Comment (dostuffthatmatters) [reply src/custom_types/query.py:47]: Maybe `.to_slugged_json()`?

**Comment by MarlonMueller** [comment]
@dostuffthatmatters Thanks for the review! I integrated your requests 👍🏼.

**Review by dostuffthatmatters — APPROVED** [review]

**Comment by dostuffthatmatters** [comment]
> @dostuffthatmatters Thanks for the review! I integrated your requests 👍🏼.

Great! I will merge this branch into the new repo tomorrow.
